reconciler/leader.go

```diff
@@ -65,10 +68,13 @@ type LeaderAware interface {
 // promotion and demotion.
 type LeaderAwareFuncs struct {
 	sync.RWMutex
-	buckets map[string]Bucket
+	buckets    map[string]Bucket
+	metricsCtx context.Context
+	once       sync.Once
 
-	PromoteFunc func(b Bucket, enq func(Bucket, types.NamespacedName)) error
-	DemoteFunc  func(b Bucket)
+	WorkQueueName string
+	PromoteFunc   func(b Bucket, enq func(Bucket, types.NamespacedName)) error
+	DemoteFunc    func(b Bucket)
```

Comment:
This (how much we have to add here) is pretty rough, and I don't really think it's worth it if we have to pollute this so much 😬

I was thinking that this logic would be in `controller.Impl`, which already has the `Name`.  Ack that the LeaderAware interface likely doesn't expose enough metadata for metrics as-is, but perhaps instead we can:
1. Expose an additional interface that `controller.Impl` can consume to emit metrics, and
2. Have `LeaderAwareFuncs` implement that interface with the metadata it has.

If that doesn't make sense, I'd be happy to jump on a quick call and we can hash it out.

Reply:
Yeah that makes sense. In that way, I think we need a timer in `controller.Impl` to report metrics periodically instead of reporting when promoting/demoting. Is that right?

Reply:
Ping.

Reply:
Maybe we should dig into https://godoc.org/k8s.io/client-go/tools/leaderelection#SetProvider ?

Reply:
https://godoc.org/k8s.io/client-go/tools/leaderelection#SwitchMetric only has a parameter for `On` and `Off` which is the  identity of the leader. Only the first call of `SetProvider` has an effect. IIUC, we'd like to count owned buckets on reconciler level instead of pod level. Unless we wrap reconciler name into the identity, we can't distinguish among reconcilers. @mattmoor WDYT?

Reply:
@mattmoor ping.

I don't think https://godoc.org/k8s.io/client-go/tools/leaderelection#SetProvider provides a way to set useful labels. Should we just implement our own logic?

Reply:
File an issue upstream?